pkg/controller/seed-controller-manager/ipam/controller.go

```diff
@@ -293,22 +291,35 @@ func (r *Reconciler) generateNewClusterAllocationForPool(ctx context.Context, cl
 
 	switch dcIPAMPoolCfg.Type {
 	case kubermaticv1.IPAMPoolAllocationTypeRange:
-		addresses, err := findFirstFreeRangesOfPool(ipamPool.Name, string(dcIPAMPoolCfg.PoolCIDR), dcIPAMPoolCfg.AllocationRange, dcIPAMPoolUsageMap)
+		ipsAllocated, err := getIPsFromAddressRanges(oldClusterAllocation.Spec.Addresses)
+		if err != nil {
+			return err
+		}
+		newIPRangeToAllocate := dcIPAMPoolCfg.AllocationRange - len(ipsAllocated)
+
+		addresses, err := findFirstFreeRangesOfPool(ipamPool.Name, string(dcIPAMPoolCfg.PoolCIDR), newIPRangeToAllocate, dcIPAMPoolUsageMap)
 		if err != nil {
 			return err
 		}
-		newClustersAllocation.Spec.Addresses = addresses
+		newClustersAllocation.Spec.Addresses = oldClusterAllocation.Spec.Addresses
+		newClustersAllocation.Spec.Addresses = append(newClustersAllocation.Spec.Addresses, addresses...)
 	case kubermaticv1.IPAMPoolAllocationTypePrefix:
-		subnetCIDR, err := findFirstFreeSubnetOfPool(ipamPool.Name, string(dcIPAMPoolCfg.PoolCIDR), dcIPAMPoolCfg.AllocationPrefix, dcIPAMPoolUsageMap)
+		subnetCIDR, err := findFirstFreeSubnetOfPool(ipamPool.Name, string(dcIPAMPoolCfg.PoolCIDR), string(oldClusterAllocation.Spec.CIDR), dcIPAMPoolCfg.AllocationPrefix, dcIPAMPoolUsageMap)
 		if err != nil {
 			return err
 		}
 		newClustersAllocation.Spec.CIDR = kubermaticv1.SubnetCIDR(subnetCIDR)
 	}
 
-	err := r.Create(ctx, newClustersAllocation)
+	obj := newClustersAllocation.DeepCopyObject().(*kubermaticv1.IPAMAllocation)
+	_, err := controllerruntime.CreateOrUpdate(ctx, r.Client, obj, func() error {
```

Comment:
This function is slightly misused, however it might still work . I think by running DeepCopyObject you already provide the reconciled object and no further changes would be needed. It is sufficient to provide the object with name and namespace  metadata only. As an example you can checkout the tests of the specific library:
https://github.com/kubernetes-sigs/controller-runtime/blob/v0.15.1/pkg/controller/controllerutil/controllerutil_test.go#L267C1-L267C32

Reply:
Good point, but with the current way how we process IPAMAllocation I needed to use a whole object to properly mutate it during creation and update. 
I think I still miss something to make it work with ObjectMeta only.

For the next iterations I would like to refactor it and seperate this logic
for creating and updating.